plugins/cache_promote/cache_promote.cc

```diff
@@ -337,10 +286,128 @@ class LRUPolicy : public PromotionPolicy
   unsigned _hits    = 10;
   // For the LRU. Note that we keep track of the List sizes, because some versions fo STL have broken
   // implementations of size(), making them obsessively slow on calling ::size().
-  TSMutex _lock;
+  TSMutex _lock{TSMutexCreate()};
   LRUMap _map;
   LRUList _list, _freelist;
   size_t _list_size = 0, _freelist_size = 0;
+
+  class Do_promote_steps
+  {
+  public:
+    Do_promote_steps(TSHttpTxn txnp, LRUPolicy *policy) : _txnp(txnp), _policy(policy) {}
+
+    void
+    operator()()
+    {
+      char *url   = nullptr;
+      int url_len = 0;
+      TSMBuffer request;
+      TSMLoc req_hdr;
+
+      if (TS_SUCCESS == TSHttpTxnClientReqGet(_txnp, &request, &req_hdr)) {
+        TSMLoc c_url = TS_NULL_MLOC;
+
+        // Get the cache key URL (for now), since this has better lookup behavior when using
+        // e.g. the cachekey plugin.
+        if (TS_SUCCESS == TSUrlCreate(request, &c_url)) {
+          if (TS_SUCCESS == TSHttpTxnCacheLookupUrlGet(_txnp, request, c_url)) {
+            url = TSUrlStringGet(request, c_url, &url_len);
+            TSHandleMLocRelease(request, TS_NULL_MLOC, c_url);
+          }
+        }
+        TSHandleMLocRelease(request, TS_NULL_MLOC, req_hdr);
+      }
+
+      // Generally shouldn't happen ...
+      if (!url) {
+        cache_turned_off(_txnp);
+
+        // Reenable and continue with the state machine.
+        TSHttpTxnReenable(_txnp, TS_EVENT_HTTP_CONTINUE);
+
+        delete this;
+
+        return;
+      }
+
+      TSDebug(PLUGIN_NAME, "LRUPolicy::doPromote(%.*s%s)", url_len > 100 ? 100 : url_len, url, url_len > 100 ? "..." : "");
+      _hash.init(url, url_len);
+      TSfree(url);
+
+      ContinueInMemberFunc<Do_promote_steps, &Do_promote_steps::critical_section>::once(*this, _policy->_lock)->schedule();
+    }
+
+  private:
+    TSHttpTxn _txnp;
+    LRUPolicy *_policy;
```

Comment:
I guess this also needs a dummy TSMutex to pass to the after_critical_section continuation.